### DISCUSSION TITLE
Test Case, Magnetic Flux Density at a distance

### DISCUSSION THREAD

**Discussion by WidderGit** [thliebig/openEMS-Project#227] [General]
Hello,
I am wondering if anyone has done a test case for the "classic" formula for calculating the magnetic flux density, B = (mu0)I/2(pi)r, for a thin wire in openEMS (preferably in Python but I'll take what I can get)?  
It is worth noting that I'm eventually interested in calculating the magnetic field generated under transient conditions (due to differential voltages at the ends of a wire), but I think the classic test case is a good start to get myself calibrated.

Thanks
Marc

**Comment by WidderGit** [comment]
Looks like there is no example available so I'll give it a try.  Except I need some help on how to get started.
I am planning to model the copper wire as a cylinder and add an input port to one side and a "ground" port to the other.
Let's say I want to apply 5V to the input port and 0V to the ground side
Can anyone tell me what type of port I should use and what the inputs are supposed to be?  

I already have follow up questions in the cue but I don't want to ask too many at one time.

Thanks,
Marc

**Reply by VolkerMuehlhaus** [reply]
> I am planning to model the copper wire as a cylinder and add an input port to one side and a "ground" port to the other.

This is not possible, current only flows in **closed** loops. In your simulation model you always need a physical path which includes a return path. That could be some plane or wire, but you must close the loop so that current can flow. 

Your idea of having a large "unphysical" port that spans from one end of the wire to the other end is not possible, such a port does not exist in openEMS. Results would be questionable anyway, because any RF port must be small compared to the wavelength.

I know that you are not happy with this answer because you want to isolate the effect of a single wire with no return, but that is not possible in most 3D RF EM simulation tools.

**Reply by WidderGit** [reply]
Thanks for the input
Do you (or anyone else) know of an openEMS transient example for a closed loop of wire (or conductor) with a voltage source/ground?
Marc

**Reply by VolkerMuehlhaus** [reply]
Hi Marc, 

I can simplify my [octagon inductor model](https://github.com/IHP-GmbH/IHP-Open-PDK/tree/main/ihp-sg13g2/libs.tech/openems/testcase/SG13_Octagon_L2n0/OpenEMS_Python) for you. But can you describe your modelling idea, why you want to use an ideal voltage source (internal resistance = 0) instead of a port with finite internal resistance?

Best regards
Volker

**Reply by WidderGit** [reply]
As a start I want to make sure I have everything set up correctly by comparing a section of the wire results against B = (mu0)I/2(pi)r.  This is assuming I will be able to find the current in the wire and the B field around it, at some consistent radial distance r.  After I verify that setup, I want to move to the transient response of parallel wires, each with a different transient input.  I use the term "wire" but I understand this can be any straight conductive path

As you know the resistance in the wires will be low, and I do not know the inductive effects yet, so I think that assuming there is a 10 Ohm resistor at the input, with a 5V input, is a good start.  If this works, I would eventually want to tune the resistance to optimize the power

Thanks for offering to help
Marc

**Reply by VolkerMuehlhaus** [reply]
It seems that I have broken my openEMS installation by installing some other Python package that required other numpy versions, so I cannot test something at the moment.

Here is a modified model, I couldn't test the changes so there might be some typos. 
The model consists of a line above ground plane, with a port on each end. You can leave this as is, or remove the second port and insert a metal block there that closes the loop.

The model uses time domain evaluation, as you requested.

Hope this helps to get you started, good luck!
Volker

edit: Updated ZIP file
[tdr_wire_v2.zip](https://github.com/user-attachments/files/16730086/tdr_wire_v2.zip)

**Reply by WidderGit** [reply]
Volker,
Thank you very much for your help.  
I am running it right now (CSXCAD came up and I think the program is still processing), looking forward to seeing the results.
I'll take a close look at the program and repost if I can't figure something out.
Good luck getting openEMS fixed.
Thanks again
Marc

**Reply by VolkerMuehlhaus** [reply]
Hi Marc,

great, hope it helps. I forgot to mention that my model doesn't store field data, it only calculates the voltage and current at the ports and plots them in time domain. I hope you can find find the code for field storage in other examples.

My openEMS installation is up and running again, so I could test the model and it works as expected. Essentially this was a TDR (time domain reflectometry) model, where we can extract discontinuities in line impedance from the step response in time domain. No field storage/field display required for that purpose, just plain voltage & current at the ports.

BR, Volker

**Reply by WidderGit** [reply]
Volker,
I'm probably doing something wrong but when I run from a CMD window, the program starts and shows the geometry in CSXCAD, but then it appears to "hang" with no feedback.  I tried hitting enter in the CMD window to see if it needed an input for the next step, but no luck.  It looks like some plots should be generated but I am not seeing them.  I did let it run for at least a half hour before checking back with the progress.

From what I can tell I should be able to add a "dump box" around the area of interest to get the B field data and output it in a Paraview file or equivalent.  Hopefully this is the case.
Thanks,
Marc

**Reply by VolkerMuehlhaus** [reply]
Hi Marc,

the simulation starts **after** you closed the CSXCAD view. So maybe you didn't close it?

BR, Volker

**Reply by WidderGit** [reply]
Volker,
Yes, that was it, I feel silly.
I am getting an assertion error that I may be able to fix so I'll not post unless I get stuck.

Thanks,
Marc

**Reply by WidderGit** [reply]
Volker,
I was able to get it to run! and have a tip for others that I found in the forum.
If you get "File "openEMS\openEMS.pyx", line 489, in openEMS.openEMS.openEMS.Run
AssertionError"
Try running "python -O tdr_wire.py" instead of "python tdr_wire.py"
Thanks
Marc

**Reply by biergaizi** [reply]
> If you get "File "openEMS\openEMS.pyx", line 489, in openEMS.openEMS.openEMS.Run
> AssertionError"

It happens if the file path contains a symbolic link, so there's a mismatch between the target path and the actual path. I thought I've fixed the assertion, but it turned out that the fix is incomplete (I've also seen this error occasionally). For now, just disable assertion with `-O`.

**Reply by WidderGit** [reply]
Thanks for explaining the source of the error
Marc
